wallet/wallet-controller/src/lib.rs

```diff
@@ -17,148 +17,281 @@
+
+        let (node_block_id, node_block_height) = match self.node_chain_info.as_ref() {
+            Some(info) => (info.best_block_id, info.best_block_height),
+            None => return,
+        };
+
+        let (wallet_block_id, wallet_block_height) =
+            self.wallet.get_best_block().expect("`get_best_block` should not fail normally");
+
+        // Wait until the node has enough block height.
+        // Block sync may not work correctly otherwise.
+        if node_block_id == wallet_block_id || node_block_height < wallet_block_height {
+            return;
+        }
+
+        let chain_config = Arc::clone(&self.chain_config);
+        let mut rpc_client = self.rpc_client.clone();
+
+        self.block_fetch_task = Some(tokio::spawn(async move {
+            let sync_res = sync::fetch_new_block(
+                &chain_config,
+                &mut rpc_client,
+                node_block_id,
+                node_block_height,
+                wallet_block_id,
+                wallet_block_height,
+            )
+            .await;
+
+            if let Err(e) = &sync_res {
+                log::error!("Block fetch failed: {e}");
+                // Wait a bit to not spam constantly if the node is unreachable
+                tokio::time::sleep(Duration::from_secs(10)).await;
+            }
+
+            sync_res
+        }));
+    }
+
+    fn handle_block_fetch_result(&mut self, res: BlockFetchResult<T>) {
+        if let Ok(FetchedBlock {
+            block,
+            block_height,
+        }) = res
+        {
+            let scan_res = self.wallet.scan_new_blocks(block_height, vec![block]);
+            if let Err(e) = scan_res {
+                log::error!("Block scan failed: {e}");
+            }
+        }
+    }
+
+    async fn recv_block_fetch_result(
+        block_fetch_task: &mut Option<JoinHandle<BlockFetchResult<T>>>,
+    ) -> BlockFetchResult<T> {
+        // This must be cancel safe!
+        match block_fetch_task {
+            Some(task) => {
+                let res = task.await.expect("Block fetch should not panic");
+                *block_fetch_task = None;
+                res
+            }
+            None => std::future::pending().await,
+        }
+    }
+
+    /// Sync the wallet block chain from the node.
+    /// This function is cancel safe.
+    pub async fn run_sync(&mut self) {
+        // This must be cancel safe!
+        loop {
+            self.start_block_fetch_if_needed();
+
+            tokio::select! {
+                chain_info_opt = self.node_state_rx.recv() => {
+                    // Channel is always open because [run_tip_sync] does not return
+                    self.handle_node_state_change(chain_info_opt.expect("Channel must be open"));
+                }
+                sync_result = Self::recv_block_fetch_result(&mut self.block_fetch_task) => {
+                    self.handle_block_fetch_result(sync_result);
+                }
+            }
+        }
     }
```

Comment:
It would be great if all these went into a separate file that can be tested in isolation...